plugin/src/main/java/org/opensearch/ml/action/mcpserver/McpAsyncServerHolder.java

```diff
@@ -22,15 +23,57 @@
  */
 public class McpAsyncServerHolder {
 
-    public static OpenSearchMcpServerTransportProvider mcpServerTransportProvider = new OpenSearchMcpServerTransportProvider(
-        new ObjectMapper()
-    );
-
     public static Map<String, StreamingRestChannel> CHANNELS = new ConcurrentHashMap<>();
+    public static Map<String, Long> IN_MEMORY_MCP_TOOLS = new ConcurrentHashMap<>();
+    private static volatile MLIndicesHandler mlIndicesHandler;
+    private static volatile McpToolsHelper mcpToolsHelper;
+
+    public static void init(MLIndicesHandler mlIndicesHandler, McpToolsHelper mcpToolsHelper) {
+        McpAsyncServerHolder.mlIndicesHandler = mlIndicesHandler;
+        McpAsyncServerHolder.mcpToolsHelper = mcpToolsHelper;
+    }
+
+    private static volatile OpenSearchMcpServerTransportProvider mcpServerTransportProvider;
+    private static volatile McpAsyncServer asyncServer;
+
+    public static OpenSearchMcpServerTransportProvider getMcpServerTransportProviderInstance() {
+        if (mcpServerTransportProvider != null) {
+            return mcpServerTransportProvider;
+        }
+        synchronized (McpAsyncServerHolder.class) {
+            if (mcpServerTransportProvider != null) {
+                return mcpServerTransportProvider;
+            }
+            mcpServerTransportProvider = new OpenSearchMcpServerTransportProvider(
+                McpAsyncServerHolder.mlIndicesHandler,
+                McpAsyncServerHolder.mcpToolsHelper,
+                new ObjectMapper()
+            );
+            if (asyncServer == null) {
+                asyncServer = McpServer
+                    .async(mcpServerTransportProvider)
+                    .capabilities(McpSchema.ServerCapabilities.builder().tools(true).logging().build())
+                    .serverInfo("OpenSearch-MCP-Server", "0.1.0")
```

Comment:
Let's add these as a static variable?

Reply:
Same above, I don't think it's necessary.